pywbem_mock/_wbemconnection_mock.py

```diff
@@ -1024,12 +1051,607 @@ def _mock_methodcall(self, methodname, localobject, Params=None, **params):
+    #  The following methods map the imethodcall interface (dictionary of
+    #  the method parameters to the arguments required by each method in the
+    #  corresponding CIM repository and call that method. the results
+    #  are mapped back to be compatible with imethodcall return tuple.
     #
     #####################################################################
 
+    # Instance Operations
+
+    def _imeth_EnumerateInstanceNames(self, **params):
+        """
+        Call :meth:`~pywbem_mock.MainProvider.EnumerateInstanceNames` with
+        parameters defined for that method and map response to tuple response
+        for imethodcall.
+
+        Parameters:
+
+          params (class:`py:dict`):
+            Dictionary of parameters for the method called.
+
+        Returns:
+
+          Tuple with instance paths comatible with imethodcall
+
+        Raises:
+
+          Error: Exceptions from the call
+        """
+        instance_paths = self.mainprovider.EnumerateInstanceNames(
+            ClassName=_cvt_rqd_classname(params['ClassName']),
+            namespace=self.namespace)
+        return self._make_tuple(instance_paths)
+
+    def _imeth_EnumerateInstances(self, **params):
+        """
+        Call :meth:`~pywbem_mock.MainProvider.EnumerateInstances` with
+        parameters defined for that method and map response to tuple response
+        for imethodcall.
+
+        Parameters:
+
+          params (class:`py:dict`):
+            Dictionary of parameters for the method called.
+
+        Returns:
+
+          Tuple with instances comatible with imethodcall
+
+        Raises:
+
+          Error: Exceptions from the call
+        """
+        instances = self.mainprovider.EnumerateInstances(
+            ClassName=_cvt_rqd_classname(params['ClassName']),
+            namespace=self.namespace,
+            LocalOnly=params.get('LocalOnly', None),
+            DeepInheritance=params.get('DeepInheritance', None),
+            IncludeQualifiers=params.get('IncludeQualifiers', None),
+            IncludeClassOrigin=params.get('IncludeClassOrigin', None),
+            PropertyList=params.get('PropertyList', None))
+        return self._make_tuple(instances)
+
+    def _imeth_GetInstance(self, **params):
+        """
+        Call :meth:`~pywbem_mock.MainProvider.GetInstance` with
+        parameters defined for that method and map response to tuple response
+        for imethodcall.
+
+        This method includes the namespace within the InstanceName
+        rather than as a separate element.
+
+        Parameters:
+
+          params (class:`py:dict`):
+
+            Dictionary of parameters for the method called.
+
+        Returns:
+          Tuple with instance comatible with imethodcall
+
+        Raises:
+
+          Error: Exceptions from the call
+        """
+        InstanceName = params['InstanceName']
+        if InstanceName.namespace is None:
+            InstanceName.namespace = self.namespace
```

Comment:
I think this comment is misleading, and the code that sets `InstanceName.namespace` a few lines down works but could be made more consistent.

The call to this method is done in `_mock_imethodcall()` which generically passes on its own arguments. In the case of the single instance operations like GetInstance, that includes a `namespace` parameter and the `InstanceName` parameter. The `WBEMConnection.GetInstance()` code ensures that the `namespace` parameter is always set (from the `InstanceName` namespace, defaulting to the default namespace of the connection), but it passes the `InstanceName` parameter on without ensuring that its namespace is set. See [here](https://github.com/pywbem/pywbem/blob/master/pywbem/_cim_operations.py#L2818).

As a result, this method here:
* gets a `namespace` parameter that is reliably set.
* gets an `InstanceName` parameter whose namespace is as originally passed in by the user who called `(Faked)WBEMConnection.GetInstance()`, i.e. it may or may not be set.

Therefore, when this method determines the namespace to be used a few lines down from here, it should in addition take the `namespace` parameter into account.

Reply:
I went wrong by assuming that self.namespace was the default namespace which it is not. Sorry for that. I agree with the latest change to pass the namespace as a parameter to the `_imeth_*()` methods.